safir/src/safir/metrics/_event_manager.py

```diff
@@ -556,3 +592,54 @@ async def build_publisher_for_model(
             An appropriate event publisher implementation instance.
         """
         return NoopEventPublisher[P](self._application, model, self.logger)
+
+
+class MockEventManager(EventManager):
+    """An event manager that creates mock publishers that record all publishes.
+
+    This is used as the implementation of `~safir.metrics.EventManager` when
+    event publication is disabled and mocking is enabled. Like a
+    `~safir.metrics.NoopEventManager`, the event type registrations are still
+    verified to catch errors, but any calls to
+    `~safir.metrics.MockEventPublisher` are recorded for later assertion.
+
+    This is for use only in unit testing. Don't use it in any deployed
+    environment because memory usage will grow unbounded.
+
+
+    Parameters
+    ----------
+    application
+        Name of the application that is generating events.
+    topic_prefix
+        Kafka topic prefix for the metrics events topic for this application.
+    logger
+        Logger to use for internal logging.
+    """
+
+    def __init__(
+        self,
+        application: str,
+        topic_prefix: str,
+        logger: BoundLogger | None = None,
+    ) -> None:
+        super().__init__(f"{topic_prefix}.{application}", logger)
+        self._application = application
+
+    async def build_publisher_for_model(
+        self, model: type[P]
+    ) -> EventPublisher[P]:
+        """Build a no-op recording publisher for a specific enriched model.
+
+        Parameters
+        ----------
+        model
+            Enriched and configured model representing the event that will be
+            published.
+
+        Returns
+        -------
+        EventPublisher
+            An appropriate event publisher implementation instance.
+        """
+        return MockEventPublisher[P](self._application, model, self.logger)
```

Comment:
With the change above to the constructor argument type, I believe you can then omit the `[P]` qualifier here and mypy will figure it out for you.